compiler_test.go

```diff
@@ -255,6 +262,127 @@ message Foo {
+
+	descriptor, err := protoregistry.GlobalFiles.FindFileByPath(descriptorProtoPath)
+	require.NoError(t, err)
+	descriptorProto := protodesc.ToFileDescriptorProto(descriptor)
+
+	// We will share this descriptor/parse result (which needs to be modified by the linker
+	// to resolve all references) from multiple concurrent operations to make sure the race
+	// detector is not triggered.
+	testCases := []struct {
+		name     string
+		resolver Resolver
+	}{
+		// TODO: Sadly, interpreting options does not actually work when no AST is provided.
+		//       Uncomment this test case when this is fixed.
+		// {
+		//	name: "share unresolved descriptor",
+		//	resolver: WithStandardImports(ResolverFunc(func(name string) (SearchResult, error) {
+		//		if name == "desc_test_complex.proto" {
+		//			return SearchResult{
+		//				Proto: parseResult.FileDescriptorProto(),
+		//			}, nil
+		//		}
+		//		return SearchResult{}, os.ErrNotExist
+		//	})),
+		// },
+		{
+			name: "share resolved descriptor",
+			resolver: WithStandardImports(ResolverFunc(func(name string) (SearchResult, error) {
+				if name == "desc_test_complex.proto" {
+					return SearchResult{
+						Proto: resolvedProto,
+					}, nil
+				}
+				return SearchResult{}, os.ErrNotExist
+			})),
+		},
+		{
+			name: "share unresolved parse result",
+			resolver: WithStandardImports(ResolverFunc(func(name string) (SearchResult, error) {
+				if name == "desc_test_complex.proto" {
+					return SearchResult{
+						ParseResult: parseResult,
+					}, nil
+				}
+				return SearchResult{}, os.ErrNotExist
+			})),
+		},
+		{
+			name: "share google/protobuf/descriptor.proto",
+			resolver: WithStandardImports(ResolverFunc(func(name string) (SearchResult, error) {
+				// we'll parse our test proto from source, but its implicit dep on
+				// descriptor.proto will use a
+				switch name {
+				case "desc_test_complex.proto":
+					return SearchResult{
+						Source: bytes.NewReader(data),
+					}, nil
+				case "google/protobuf/descriptor.proto":
+					return SearchResult{
+						Proto: descriptorProto,
+					}, nil
+				default:
+					return SearchResult{}, os.ErrNotExist
+				}
+			})),
+		},
+	}
+
+	for i := range testCases {
+		testCase := testCases[i]
+		t.Run(testCase.name, func(t *testing.T) {
+			t.Parallel()
+			compiler1 := &Compiler{
+				Resolver:       testCase.resolver,
+				SourceInfoMode: SourceInfoStandard,
+			}
+			compiler2 := &Compiler{
+				Resolver:       testCase.resolver,
+				SourceInfoMode: SourceInfoStandard,
+			}
+			grp, ctx := errgroup.WithContext(context.Background())
+			grp.Go(func() error {
+				_, err := compiler1.Compile(ctx, "desc_test_complex.proto")
+				return err
+			})
+			grp.Go(func() error {
+				// We need to start this *after* the one above, but we can't
+				// use any sychronizing event or that would not be a race.
+				// So we assume a one second delay is sufficient.
+				time.Sleep(time.Second)
```

Comment:
Can't you use a sync.WaitGroup or something here to avoid the sleep? Do you want them to run one first then the other, or do you want the two goroutines to be running at the same time?

Reply:
See the above comment -- a `WaitGroup` would be a synchronizing event that would remove the data race. The race detector only complains when a read or write in one goroutine happens after a write in another, but _without_ any synchronization events (mutexes, channels, wait groups, etc).

We need the writes to happen first, and _then_ the reads. But since we _can't_ use a synchronization mechanism, we have to just sleep to ensure the desired order of operations. And, yes, it's not deterministic -- they could possibly be concurrent, like on a really slow CI container. But in practice, it should be fine. (On my MBP, the actual compile only takes like ~20 milliseconds, so the 1 second sleep is way more than sufficient.)

Reply:
Maybe I'm missing something, but wouldn't this guarantee that the first goroutine compiled before the second?

```diff
diff --git a/compiler_test.go b/compiler_test.go
index eb13dd9..a531f85 100644
--- a/compiler_test.go
+++ b/compiler_test.go
@@ -20,8 +20,8 @@ import (
 	"errors"
 	"os"
 	"strings"
+	"sync"
 	"testing"
-	"time"
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
@@ -365,15 +365,16 @@ func TestDataRace(t *testing.T) {
 				SourceInfoMode: SourceInfoStandard,
 			}
 			grp, ctx := errgroup.WithContext(context.Background())
+			var wg sync.WaitGroup
+			wg.Add(1)
 			grp.Go(func() error {
+				defer wg.Done()
 				_, err := compiler1.Compile(ctx, "desc_test_complex.proto")
 				return err
 			})
 			grp.Go(func() error {
-				// We need to start this *after* the one above, but we can't
-				// use any sychronizing event or that would not be a race.
-				// So we assume a one second delay is sufficient.
-				time.Sleep(time.Second)
+				// We need to start this *after* the one above.
+				wg.Wait()
 				_, err := compiler2.Compile(ctx, "desc_test_complex.proto")
 				return err
 			})
```

Reply:
n/m - I re-read your comment and understand now why this wouldn't work.